src/uart/mod.rs

```diff
@@ -0,0 +1,81 @@
+//! Using the UART
+use crate::error::{NegativeErrorExt, NumericError};
+use embedded_io;
+use embedded_io::ErrorKind;
+use riot_sys::libc::c_void;
+use riot_sys::uart_init;
+use riot_sys::uart_t;
+use riot_sys::uart_write;
+
+#[derive(Debug)]
+pub struct UARTDevice {
+    dev: uart_t,
+}
+
+#[derive(Debug)]
+pub struct Error(NumericError);
+
+impl From<NumericError> for Error {
+    fn from(err: NumericError) -> Self {
+        Self(err)
+    }
+}
+
+impl embedded_io::Error for Error {
+    fn kind(&self) -> ErrorKind {
+        match -self.0.number() as _ {
+            riot_sys::ENODEV => ErrorKind::NotFound,
+            riot_sys::ENOTSUP => ErrorKind::Unsupported,
+            _ => ErrorKind::Other,
+        }
+    }
+}
+
+impl embedded_io::ErrorType for UARTDevice {
+    type Error = Error;
+}
+
+impl UARTDevice {
+    /// Create a new UARTDevice from a RIOT descriptor
+    pub fn new(dev: uart_t) -> Self {
+        UARTDevice { dev }
+    }
+
+    pub fn init(
+        &mut self,
+        baudrate: u32,
+        callback: Option<unsafe extern "C" fn(ctx: *mut c_void, byte: u8)>,
+        arg: Option<*mut c_void>,
+    ) -> Result<(), Error> {
+        let result = {
+            unsafe {
+                uart_init(
+                    self.dev,
+                    baudrate,
+                    callback,
+                    arg.unwrap_or(core::ptr::null_mut()),
+                )
+            }
+        };
+
+        match result.negative_to_error() {
+            Ok(_) => Ok(()),
+            Err(err) => Err(Error::from(err)),
+        }
```

Comment:
```suggestion
        match result.negative_to_error() {
            Ok(_) => Ok(()),
            Err(err) => Err(Error::from(err)),
        }
```
There got to be a better way...

Reply:
The prettier way would be `Ok(result.negative_to_error()?)`, because `?` implicitly calls `.into()` on the result. If there were no implicit `.into()`, there'd also be `.map_err(Error::from)` or `.map_err(Error)`. If we go with embedded_io::Error for NumericError, that's all moot, of course.